### PR TITLE
Add minimum version to options, needed to fix docker/docker#30258

### PR DESCRIPTION
**Pull request by adshmh**
This is an attempt to add the code needed for fixing [docker/docker#30258](https://github.com/docker/docker/issues/30258). 

If the PR is approved, I can add code to docker/docker for passing the minimum TLS version to go-connections. 

Signed-off-by: Arash Deshmeh <adeshmeh@ca.ibm.com>

### CODE REVIEW

**Comment by adshmh** [comment]
@tiborvass @cyli

File: tlsconfig/config.go [hunk 1]
```diff
@@ -83,6 +84,22 @@ func certPool(caFile string) (*x509.CertPool, error) {
 	return certPool, nil
 }
 
+// minVersion returns the minimum TLS version matching 'version', the MinVersion string.
+func minVersion(version string) (uint16, error) {
+	allVersions := map[string]uint16{
+		"VersionSSL30": tls.VersionSSL30,
```
Comment (tiborvass) [inline tlsconfig/config.go]: @cyli I am not convinced we should reuse the same string as the Go constants. If anything I would remove the `Version` prefix but `SSL30`, `TLS10`, `TLS11` and `TLS12` do not seem to be something very widespread. Maybe `SSL3.0`, or `SSL 3.0` ? I couldn't find in the RFC any standard string for this.

Comment (tiborvass) [inline tlsconfig/config.go]: Think of this as, this would be the string displayed in the docker config at `"MinVersion": "VersionSSL30"`

Comment (cyli) [inline tlsconfig/config.go]: I'm not sure that the string translation should happen in the `go-connections` library.  Agree that there doesn't seem to be any standard.  Probably `go-connections` should just take a `uint16` like `tls.Config`, and enforce that that number is at least >= `tls.VersionTLS10`?

The user-friendly translation can happen at the user-facing layer (like a CLI or config file parser).

Comment (tiborvass) [inline tlsconfig/config.go]: Agreed. @adshmh would you mind removing this function from here and pass in a `uint16` instead of string?

File: tlsconfig/config.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ type Options struct {
 	InsecureSkipVerify bool
 	// server-only option
 	ClientAuth tls.ClientAuthType
+	MinVersion string
```
Comment (tiborvass) [inline tlsconfig/config.go]: Basically making this a `uint16`.

File: tlsconfig/config.go (hunk 1 above)
Comment (adshmh) [inline tlsconfig/config.go]: @tiborvass and @cyli Thank you for the review. I will update the PR.

Comment (adshmh) [inline tlsconfig/config.go]: @tiborvass  I changed the type and removed the conversion function but kept a basic function to check the input validity.

**Comment by adshmh** [comment]
@cyli Thank you for pointing out that issue. I have added the code required to include the client.

**Comment by cyli** [comment]
@adshmh Thanks for adding the client code and all the tests!  LGTM

**Comment by cyli** [comment]
@adshmh Apologies, would you mind rebasing?

**Comment by adshmh** [comment]
@cyli sure. Just rebased the PR.

**Comment by cyli** [comment]
Thanks kindly!  :)

File: tlsconfig/config.go [hunk 3]
```diff
@@ -96,6 +96,40 @@ func certPool(caFile string, exclusivePool bool) (*x509.CertPool, error) {
...
+// It assumes the current MinVersion on the `config` is the lowest allowed.
+func adjustMinVersion(options Options, config *tls.Config) error {
+	if options.MinVersion > 0 {
+		if !isValidMinVersion(options.MinVersion) {
+			return fmt.Errorf("Invalid minimum TLS version: %x", options.MinVersion)
+		}
+		if options.MinVersion >= config.MinVersion {
+			config.MinVersion = options.MinVersion
+		} else {
+			return fmt.Errorf("Requested minimum TLS version is too low.")
```
Comment (tiborvass) [inline tlsconfig/config.go]: Could be useful to include `config.MinVersion` in the error.

File: tlsconfig/config.go [hunk 4]
```diff
@@ -96,6 +96,40 @@ func certPool(caFile string, exclusivePool bool) (*x509.CertPool, error) {
 	return certPool, nil
 }
 
+// isValidMinVersion checks that the input value is a valid tls minimum version
+func isValidMinVersion(version uint16) bool {
+	allVersions := []uint16{
```
Comment (tiborvass) [inline tlsconfig/config.go]: this could be a map for an O(1) lookup and could be a package-level variable.

File: tlsconfig/config.go (hunk 3 above)
Comment (tiborvass) [inline tlsconfig/config.go]: Also not too important, but it's slightly better to `if options.MinVersion < config.MinVersion { return fmt.Errorf(...) }`, it won't require any `else` statement.

File: tlsconfig/config_test.go [hunk 5]
```diff
@@ -377,6 +377,70 @@ func TestConfigServerExclusiveRootPools(t *testing.T) {
...
+// If a valid minimum version is specified in the options, the server's
+// minimum version should be set accordingly
+func TestConfigServerTLSMinVersionIsSetBasedOnOptions(t *testing.T) {
+	versions := []uint16{
+		tls.VersionTLS11,
+		tls.VersionTLS12,
+	}
+	tempDir := makeTempDir(t)
+	defer os.RemoveAll(tempDir)
+	key, cert := generateCertAndKey(t, tempDir)
```
Comment (tiborvass) [inline tlsconfig/config_test.go]: Just a note for myself or whoever wants to fix this, but we should be using generated certs/keys for tests and not regenerate every time. Should be done in another PR.

Comment (adshmh) [inline tlsconfig/config_test.go]: #dibs. I can submit a PR for this after the current one is done.

File: tlsconfig/config.go [hunk 6]
```diff
@@ -96,6 +105,30 @@ func certPool(caFile string, exclusivePool bool) (*x509.CertPool, error) {
 	return certPool, nil
 }
 
+// isValidMinVersion checks that the input value is a valid tls minimum version
+func isValidMinVersion(version uint16) bool {
+	if _, ok := allTLSVersions[version]; !ok {
```
Comment (tiborvass) [inline tlsconfig/config.go]: or simply `return ok` :)

File: tlsconfig/config_test.go (hunk 5 above)
Comment (tiborvass) [inline tlsconfig/config_test.go]: Thanks!

**Comment by tiborvass** [comment]
LGTM
